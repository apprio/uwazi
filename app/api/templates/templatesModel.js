import mongoose from 'mongoose';

const propertiesSchema = new mongoose.Schema({
  id: String,
  label: String,
  type: String,
  content: String,
  relationType: String,
  name: String,
  filter: Boolean,
  noLabel: Boolean,
  fullWidth: Boolean,
  defaultfilter: Boolean,
  required: Boolean,
  sortable: Boolean,
  showInCard: Boolean,
  prioritySorting: Boolean,
  style: String,
  nestedProperties: [String]
});

const commonPropertiesSchema = new mongoose.Schema({
  isCommonProperty: Boolean,
  label: String,
  name: String,
  type: String,
  prioritySorting: Boolean
});

const templateSchema = new mongoose.Schema({
  name: String,
  default: Boolean,
  properties: [propertiesSchema],
  commonProperties: [commonPropertiesSchema]
});

const Model = mongoose.model('templates', templateSchema);
export default Model;
