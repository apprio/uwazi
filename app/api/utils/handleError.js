/* eslint-disable prefer-template */
import debugLog from 'api/log/debugLog';
import errorLog from 'api/log/errorLog';

export default (error, { req = {}, uncaught = false } = {}) => {
  let result = error;
  const responseToClientError = error.json;

  if (responseToClientError) {
    return false;
  }

  if (error instanceof Error) {
    result = { code: 500, message: error.stack };
  }

  if (error.name === 'MongoError') {
    result.code = 500;
  }

  if (error.message.match(/Cast to ObjectId failed for value/)) {
    result.code = 400;
  }

  if (error.message.match(/rison decoder error/)) {
    result.code = 400;
  }

  if (uncaught) {
    result.message = `uncaught exception or unhandled rejection, Node process finished !!\n ${result.message}`;
  }


  const errorMessage = (req.originalUrl ? `\nurl: ${req.originalUrl}` : '') +
    (req.body && Object.keys(req.body).length ? `\nbody: ${JSON.stringify(req.body, null, ' ')}` : '') +
    (req.query && Object.keys(req.query).length ? `\nquery: ${JSON.stringify(req.query, null, ' ')}` : '') +
    `\n${result.message}`;

  if (result.code === 500) {
    errorLog.error(errorMessage);
  }

  if (result.code === 400) {
    debugLog.debug(errorMessage);
  }

  return result;
};
