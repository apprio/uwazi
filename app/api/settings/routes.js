import Joi from 'joi';
import settings from 'api/settings/settings';
import { validateRequest } from '../utils';
import needsAuthorization from '../auth/authMiddleware';

export default (app) => {
  app.post('/api/settings',
    needsAuthorization(),
    validateRequest(Joi.object().keys({
      _id: Joi.string(),
      __v: Joi.number(),
      project: Joi.string(),
      site_name: Joi.string().allow(''),
      home_page: Joi.string().allow(''),
      private: Joi.boolean(),
      mailerConfig: Joi.string().allow(''),
      analyticsTrackingId: Joi.string().allow(''),
      matomoConfig: Joi.string().allow(''),
      dateFormat: Joi.string().allow(''),
      custom: Joi.any(),
      customCSS: Joi.string().allow(''),
      languages: Joi.array().items(
        Joi.object().keys({
          _id: Joi.string(),
          key: Joi.string(),
          label: Joi.string(),
          default: Joi.boolean()
        })
      ),
      filters: Joi.array().items(
        Joi.object().keys({
          _id: Joi.string(),
          id: Joi.string(),
          name: Joi.string(),
          items: Joi.any()
        })
      ),
      links: Joi.array().items(
        Joi.object().keys({
          _id: Joi.string(),
          localID: Joi.string(),
          title: Joi.string(),
          url: Joi.string()
        })
      )
    }).required()),
    (req, res, next) => {
      settings.save(req.body)
      .then(response => res.json(response))
      .catch(next);
    }
  );

  app.get('/api/settings', (req, res, next) => {
    settings.get()
    .then(response => res.json(response))
    .catch(next);
  });
};
