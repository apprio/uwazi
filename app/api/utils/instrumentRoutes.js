import * as utils from 'api/utils';

const createSpy = (key, resolve) => jasmine.createSpy(key).and.callFake((...args) => {
  resolve(`${key}:${args.join(',')}`);
});

const executeRoute = (method, routePath, req = {}, res, next = () => {}, app, runRoute = true) => {
  const args = app[method].calls.allArgs().find(a => a[0] === routePath);
  if (!args) {
    throw new Error(`Route ${method.toUpperCase()} ${routePath} is not defined`);
  }
  const result = new Promise((resolve, reject) => {
    let statusCode;

    res.status = (code) => {
      statusCode = code;
    };

    res.download = createSpy('download', resolve);
    res.redirect = createSpy('redirect', resolve);
    res.sendFile = createSpy('sendFile', resolve);

    res.json = jasmine.createSpy('json').and.callFake((response) => {
      if (statusCode) {
        response.status = statusCode;
      }
      resolve(response);
    });

    if (!args[1]) {
      return reject(new Error('route function has not been defined !'));
    }

    if (!runRoute) {
      return resolve();
    }

    const mockedNext = (error) => {
      next(error);
      reject(error);
    };

    return args[args.length - 1](req, res, mockedNext);
  });

  if (args) {
    result.middlewares = args.slice(1, -1).filter(arg => !arg.isJoi);
    const validation = args.slice(1, -1).find(arg => arg.isJoi);
    if (validation) {
      result.validation = validation.describe();
    }
  }

  return result;
};

export default (route, io) => {
  const app = jasmine.createSpyObj('app', ['get', 'post', 'delete', 'use']);
  const originalValidateRequest = utils.validateRequest;
  spyOn(utils, 'validateRequest').and.callFake(schema => schema);
  route(app, io);
  utils.validateRequest = originalValidateRequest;

  const get = (routePath, req, res = {}, next) => executeRoute('get', routePath, req, res, next, app);
  const _get = (routePath, req, res = {}, next) => executeRoute('get', routePath, req, res, next, app, false);
  get.validation = (routePath, req, res = {}, next) => executeRoute('get', routePath, req, res, next, app, false).validation;

  const post = (routePath, req, res = {}, next) => executeRoute('post', routePath, req, res, next, app);
  const _post = (routePath, req, res = {}, next) => executeRoute('post', routePath, req, res, next, app, false);
  post.validation = (routePath, req, res = {}, next) => executeRoute('post', routePath, req, res, next, app, false).validation;

  const remove = (routePath, req, res = {}, next) => executeRoute('delete', routePath, req, res, next, app);
  const _remove = (routePath, req, res = {}, next) => executeRoute('delete', routePath, req, res, next, app, false);
  remove.validation = (routePath, req, res = {}, next) => executeRoute('delete', routePath, req, res, next, app, false).validation;

  const instrumentedRoute = {
    get,
    _get,
    delete: remove,
    _delete: _remove,
    post,
    _post,
  };

  return instrumentedRoute;
};
