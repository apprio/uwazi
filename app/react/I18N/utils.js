import * as Cookie from 'tiny-cookie';

import { isClient } from 'app/utils';

const I18NUtils = {
  getLocaleIfExists: (locale, languages = []) => {
    if (!locale) {
      return null;
    }

    if (languages.find(l => l.key === locale)) {
      return locale;
    }

    return I18NUtils.getDefaultLocale(languages);
  },

  getCoockieLocale: (cookie = {}, languages) => {
    if (isClient && Cookie.get('locale')) {
      return I18NUtils.getLocaleIfExists(Cookie.get('locale'), languages);
    }

    return I18NUtils.getLocaleIfExists(cookie.locale, languages);
  },

  getDefaultLocale: languages => (languages.find(language => language.default) || {}).key,

  getLocale: (urlLanguage, languages = [], cookie = {}) => I18NUtils.getLocaleIfExists(urlLanguage, languages)
  || I18NUtils.getCoockieLocale(cookie, languages)
  || I18NUtils.getDefaultLocale(languages),

  saveLocale: (locale) => {
    if (isClient) {
      Cookie.set('locale', locale, { expires: 365 * 10 });
    }
  }
};

export default I18NUtils;
