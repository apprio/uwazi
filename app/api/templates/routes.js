import Joi from 'joi';

import settings from 'api/settings';

import { validateRequest } from '../utils';
import needsAuthorization from '../auth/authMiddleware';
import templates from './templates';

export default (app) => {
  app.post('/api/templates',
    needsAuthorization(),
    validateRequest(Joi.object().keys({
      _id: Joi.string(),
      __v: Joi.number(),
      name: Joi.string().required(),
      isEntity: Joi.boolean(),
      properties: Joi.array().required().items(
        Joi.object().keys({
          _id: Joi.string(),
          id: Joi.string(),
          localID: Joi.string(),
          label: Joi.string(),
          name: Joi.string(),
          nestedProperties: Joi.array(),
          type: Joi.string(),
          relationType: Joi.string(),
          filter: Joi.boolean(),
          noLabel: Joi.boolean(),
          defaultfilter: Joi.boolean(),
          required: Joi.boolean(),
          sortable: Joi.boolean(),
          showInCard: Joi.boolean(),
          fullWidth: Joi.boolean(),
          content: Joi.string().allow(['']),
          prioritySorting: Joi.boolean(),
          style: Joi.string(),
          inserting: Joi.any()
        })
      ),
      commonProperties: Joi.array().items(
        Joi.object().keys({
          _id: Joi.string(),
          localID: Joi.string(),
          isCommonProperty: Joi.boolean(),
          label: Joi.string(),
          name: Joi.string(),
          prioritySorting: Joi.boolean(),
          type: Joi.string()
        })
      )
    }).required()),
    (req, res, next) => {
      templates.save(req.body, req.language)
      .then((response) => {
        res.json(response);
        req.io.sockets.emit('templateChange', response);
      })
      .catch(next);
    }
  );

  app.get('/api/templates', (req, res, next) => {
    templates.get()
    .then(response => res.json({ rows: response }))
    .catch(next);
  });

  app.delete('/api/templates',
  needsAuthorization(),
  validateRequest(Joi.object({
    _id: Joi.string().required()
  }).required(), 'query'),
    (req, res, next) => {
      const template = { _id: req.query._id };
      templates.delete(template)
      .then(() => settings.removeTemplateFromFilters(template._id))
      .then((newSettings) => {
        res.json(template);
        req.io.sockets.emit('updateSettings', newSettings);
        req.io.sockets.emit('templateDelete', template);
      })
      .catch(next);
    }
  );

  app.get('/api/templates/count_by_thesauri',

    validateRequest(Joi.object().keys({
      _id: Joi.string().required()
    }).required(), 'query'),

    (req, res, next) => {
      templates.countByThesauri(req.query._id)
      .then(response => res.json(response))
      .catch(next);
    }
  );
};
