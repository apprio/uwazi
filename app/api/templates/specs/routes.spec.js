import { catchErrors } from 'api/utils/jasmineHelpers';
import instrumentRoutes from 'api/utils/instrumentRoutes';
import settings from 'api/settings/settings';
import templates from '../templates';
import templateRoutes from '../routes.js';

describe('templates routes', () => {
  let routes;

  beforeEach(() => {
    routes = instrumentRoutes(templateRoutes);
  });

  describe('GET', () => {
    it('should return all templates by default', (done) => {
      spyOn(templates, 'get').and.returnValue(Promise.resolve('templates'));
      routes.get('/api/templates')
      .then((response) => {
        const docs = response.rows;
        expect(docs).toBe('templates');
        done();
      })
      .catch(catchErrors(done));
    });

    describe('when there is an error', () => {
      it('should return the error in the response', async () => {
        spyOn(templates, 'get').and.returnValue(Promise.reject(new Error('error')));
        try {
          await routes.get('/api/templates');
        } catch (error) {
          expect(error).toEqual(new Error('error'));
        }
      });
    });
  });

  describe('DELETE', () => {
    it('should have a validation schema', () => {
      expect(routes.delete.validation('/api/templates')).toMatchSnapshot();
    });
    it('should delete a template', (done) => {
      spyOn(templates, 'delete').and.returnValue(Promise.resolve('ok'));
      spyOn(settings, 'removeTemplateFromFilters').and.returnValue(Promise.resolve());
      routes.delete('/api/templates', { query: { _id: '123' } })
      .then((response) => {
        expect(templates.delete).toHaveBeenCalledWith({ _id: '123' });
        expect(response).toEqual({ _id: '123' });
        done();
      })
      .catch(catchErrors(done));
    });

    describe('when there is an error', () => {
      it('should return the error in the response', async () => {
        spyOn(templates, 'delete').and.returnValue(Promise.reject(new Error('error')));

        try {
          await routes.delete('/api/templates', { query: {} });
        } catch (error) {
          expect(error).toEqual(new Error('error'));
        }
      });
    });
  });

  describe('POST', () => {
    it('should have a validation schema', () => {
      expect(routes.post.validation('/api/templates')).toMatchSnapshot();
    });
    it('should create a template', (done) => {
      spyOn(templates, 'save').and.returnValue(new Promise(resolve => resolve('response')));
      const req = { body: { name: 'created_template', properties: [{ label: 'fieldLabel' }] }, language: 'en' };

      routes.post('/api/templates', req)
      .then((response) => {
        expect(response).toBe('response');
        expect(templates.save).toHaveBeenCalledWith(req.body, req.language);
        done();
      })
      .catch(catchErrors(done));
    });

    describe('when there is an error', () => {
      it('should return the error in the response', async () => {
        spyOn(templates, 'save').and.returnValue(Promise.reject(new Error('error')));

        try {
          await routes.post('/api/templates');
        } catch (error) {
          expect(error).toEqual(new Error('error'));
        }
      });
    });
  });

  describe('/templates/count_by_thesauri', () => {
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/templates/count_by_thesauri')).toMatchSnapshot();
    });
    it('should return the number of templates using a thesauri', (done) => {
      spyOn(templates, 'countByThesauri').and.returnValue(Promise.resolve(2));
      const req = { query: { _id: 'abc1' } };
      routes.get('/api/templates/count_by_thesauri', req)
      .then((result) => {
        expect(result).toBe(2);
        expect(templates.countByThesauri).toHaveBeenCalledWith('abc1');
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('/api/templates/setasdefault', () => {
    it('should have a validation schema', () => {
      expect(routes.post.validation('/api/templates/setasdefault')).toMatchSnapshot();
    });

    it('should call templates to set the new default', (done) => {
      spyOn(templates, 'setAsDefault').and.returnValue(Promise.resolve([{ name: 'newDefault' }, { name: 'oldDefault' }]));
      const emit = jasmine.createSpy('emit');
      const req = { body: { _id: 'abc1' }, io: { sockets: { emit } } };
      routes.post('/api/templates/setasdefault', req)
      .then((result) => {
        expect(result).toEqual({ name: 'newDefault' });
        expect(templates.setAsDefault).toHaveBeenCalledWith('abc1');
        expect(emit).toHaveBeenCalledWith('templateChange', { name: 'newDefault' });
        expect(emit).toHaveBeenCalledWith('templateChange', { name: 'oldDefault' });
        done();
      })
      .catch(catchErrors(done));
    });
  });
});
