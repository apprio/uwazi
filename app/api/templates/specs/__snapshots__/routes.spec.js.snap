// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates routes /api/templates/setasdefault should have a validation schema 1`] = `
Object {
  "children": Object {
    "_id": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`templates routes /templates/count_by_thesauri should have a validation schema 1`] = `
Object {
  "children": Object {
    "_id": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`templates routes DELETE should have a validation schema 1`] = `
Object {
  "children": Object {
    "_id": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`templates routes POST should have a validation schema 1`] = `
Object {
  "children": Object {
    "__v": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "type": "number",
    },
    "_id": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "commonProperties": Object {
      "flags": Object {
        "sparse": false,
      },
      "items": Array [
        Object {
          "children": Object {
            "_id": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "isCommonProperty": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "label": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "localID": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "name": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "prioritySorting": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "type": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
          },
          "type": "object",
        },
      ],
      "type": "array",
    },
    "default": Object {
      "falsy": Array [
        false,
      ],
      "flags": Object {
        "insensitive": true,
      },
      "truthy": Array [
        true,
      ],
      "type": "boolean",
    },
    "name": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "properties": Object {
      "flags": Object {
        "presence": "required",
        "sparse": false,
      },
      "items": Array [
        Object {
          "children": Object {
            "_id": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "content": Object {
              "type": "string",
              "valids": Array [
                "",
              ],
            },
            "defaultfilter": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "filter": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "fullWidth": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "id": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "inserting": Object {
              "type": "any",
            },
            "label": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "localID": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "name": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "nestedProperties": Object {
              "flags": Object {
                "sparse": false,
              },
              "type": "array",
            },
            "noLabel": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "prioritySorting": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "relationType": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "required": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "showInCard": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "sortable": Object {
              "falsy": Array [
                false,
              ],
              "flags": Object {
                "insensitive": true,
              },
              "truthy": Array [
                true,
              ],
              "type": "boolean",
            },
            "style": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "type": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
          },
          "type": "object",
        },
      ],
      "type": "array",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;
